orttraining/orttraining/python/training/torchdynamo/ort_backend.py

```diff
@@ -0,0 +1,519 @@
+
+
+class OrtBackend:
+    """A backend compiles (sub-)graphs in torch.fx.GraphModule to onnxruntime.InferenceSession calls.
+
+    The compiler entry point is OrtBackend.compile, which
+        1. partitions the original graph into supported sub-graphs (type: torch.fx.GrpahModule) and unsupported
+           sub-graphs.
+        2. For each supported sub-graph, it replaces its _wrapped_call function with _ort_accelerated_call.
+        3. Inside _ort_accelerated_call, it creates onnxruntime.InferenceSession and calls it to execute the sub-graph.
+    """
+
+    def __init__(self):
+        self.supported_ops = OrtOperatorSupport()
+        # TODO: this is a naive implementation of cache without proper guard
+        self.partitioner_cache: Dict[torch.fx.GraphModule, torch.fx.GraphModule] = {}
+        # TODO: this is a naive implementation of cache without proper guard, this will only work for identical inputs
+        self.ort_execution_info = OrtExecutionInfo()
+
+    def _ort_acclerated_call(self, graph_module: torch.fx.GraphModule, *args, **kwargs):
+        if graph_module in self.ort_execution_info.sessions:
+            # We have seen this graph before, so we can use cached objects including session.
+            onnx_session = self.ort_execution_info.sessions[graph_module]
+            input_names = self.ort_execution_info.input_names[graph_module]
+            output_names = self.ort_execution_info.output_names[graph_module]
+            input_devices = self.ort_execution_info.input_devices[graph_module]
+            output_devices = self.ort_execution_info.output_devices[graph_module]
+            prim_outputs = self.ort_execution_info.example_outputs[graph_module]
+        else:
+            # It's first time seeing such as graph. Let's make a new session
+            # (type: onnxruntime.InferenceSession) for it.
+
+            # TODO(wechi): this is a workaround for #84311 in PyTorch.
+            _move_placeholder_to_front(graph_module)
+            # Generate reference outputs. They are used to indicate output
+            # tensors' types and devices when calling ORT.
+            #
+            # WARNING: The downstream code should not change prim_outputs and
+            # this backend should always produces output with schema identical to prim_outputs'.
+            prim_outputs = FakeTensorProp(graph_module).propagate(*args, **kwargs)
+            self.ort_execution_info.example_outputs[graph_module] = prim_outputs
+            # Compile the torch.fx.GraphModule into a torch.jit.ScriptModule.
+            script_module = _fx_to_torchscript(graph_module)
+            # Post-processing step to add expected input and output type information
+            # to the graph in torch.jit.ScriptModule. Expected inputs is "args" and "kwargs"
+            # while expected outputs is "prim_outputs".
+            if isinstance(prim_outputs, tuple):
```

Comment:
Will it only be tuple, or will it be a list as well? Just to make sure

Reply:
`tuple` only.